sound/soc/sof/intel/hda.c

```diff
@@ -53,67 +53,114 @@ static int sdw_clock_stop_quirks = SDW_INTEL_CLK_STOP_BUS_RESET;
 module_param(sdw_clock_stop_quirks, int, 0444);
 MODULE_PARM_DESC(sdw_clock_stop_quirks, "SOF SoundWire clock stop quirks");
 
-static int sdw_params_stream(struct device *dev,
-			     struct sdw_intel_stream_params_data *params_data)
+int hda_sdw_dai_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_dai *sof_dai,
+			     struct sof_ipc_dai_config *config, bool setup, int dir)
 {
-	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
-	struct snd_soc_dai *d = params_data->dai;
-	struct sof_ipc_dai_config config;
+	struct snd_sof_widget *swidget;
 	struct sof_ipc_reply reply;
-	int link_id = params_data->link_id;
-	int alh_stream_id = params_data->alh_stream_id;
+	bool found = false;
 	int ret;
-	u32 size = sizeof(config);
-
-	memset(&config, 0, size);
-	config.hdr.size = size;
-	config.hdr.cmd = SOF_IPC_GLB_DAI_MSG | SOF_IPC_DAI_CONFIG;
-	config.type = SOF_DAI_INTEL_ALH;
-	config.dai_index = (link_id << 8) | (d->id);
-	config.alh.stream_id = alh_stream_id;
-
-	/* send message to DSP */
-	ret = sof_ipc_tx_message(sdev->ipc,
-				 config.hdr.cmd, &config, size, &reply,
+
+	/* find snd_sof_widget of type DAI with matching sname as that of snd_sof_dai name */
+	list_for_each_entry(swidget, &sdev->widget_list, list) {
+		if (dir == SNDRV_PCM_STREAM_PLAYBACK && swidget->id != snd_soc_dapm_dai_in)
+			continue;
+
+		if (dir == SNDRV_PCM_STREAM_CAPTURE && swidget->id != snd_soc_dapm_dai_out)
+			continue;
+
+		if (!strcmp(sof_dai->name, swidget->widget->sname)) {
+			found = true;
+			break;
+		}
+	}
+
+	if (!found) {
+		dev_err(sdev->dev, "error: failed to find DAI widget for: %s\n",
+			sof_dai->cpu_dai_name);
+		return -EINVAL;
+	}
+
+	/*
+	 * For static pipelines, the DAI widget would already be set up and calling
+	 * sof_widget_setup()/free() would simply return without doing anything.
+	 * For dynamic pipelines, the DAI widget would be set up or freed here.
+	 */
+	if (setup) {
+		/* set up widget first */
+		ret = sof_widget_setup(sdev, swidget);
+
+		/* and then send DAI_CONFIG IPC */
+		return sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size,
+					  &reply, sizeof(reply));
+	}
+
+	/* send DAI_CONFIG IPC first */
+	ret = sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size, &reply,
 				 sizeof(reply));
-	if (ret < 0) {
-		dev_err(sdev->dev,
-			"error: failed to set DAI hw_params for link %d dai->id %d ALH %d\n",
-			link_id, d->id, alh_stream_id);
+	if (ret < 0)
+		return ret;
+
+	/* and then free the widget */
+	return sof_widget_free(sdev, swidget);
```

Comment:
This is intended, right? In a `*_setup()` method in success case the last action is freeing the widget, and the return status of that operation is our return code. If that's indeed intended and correct, I think a comment is needed here.

Reply:
Ok, after some more reading I understand, that some "*_setup()" functions can now do either set up / initialisation or freeing depending on their `setup` parameter. But `sof_widget_setup()` is only for setup / init. Seems a bit confusing to me... Maybe the naming could be improved, but no, I have no good ideas, sorry. Maybe something along the lines of configure / reconfigure / update / amend...

Reply:
I've added a comment for the function to call this out but left the name unchanged.